glue/core/roi.py

```diff
@@ -561,6 +575,44 @@ def move_to(self, xdelta, ydelta):
         self.vy = list(map(lambda y: y + ydelta, self.vy))
 
 
+class PolygonalProjected3dROI(PolygonalROI):
+
+    """"
+    A region of interest defined by a 2d polygon in screen coordinates, as
+    defined by the projection matrix. The projection matrix should convert
+    homogeneous coordinates (x, y, z, w), where w is implicitly 1, to
+    homogeneous screen coordinates (usually the product of the world and
+    projection matrix).
+    """
+    def __init__(self, vx=None, vy=None, projection_matrix=None):
+        super(PolygonalProjected3dROI, self).__init__(vx, vy)
+        self.projection_matrix = np.asarray(projection_matrix)
+
+    def contains3d(self, x, y, z):
+        if not self.defined():
+            raise UndefinedROI
+        if not isinstance(x, np.ndarray):
+            x = np.asarray(x)
+        if not isinstance(y, np.ndarray):
+            y = np.asarray(y)
+        # work in homogeneous coordinates so we can support perspective
+        # projections as well
+        vertices = np.array([x, y, z, np.ones(len(x))])
+        # homogeneous screen coordinates
+        screen_h = np.array(np.dot(self.projection_matrix, vertices))
+        # convert to screen coordinates, and we don't care about z
+        x, y = screen_h[:2] / screen_h[3]
+        return self.contains(x, y)
```

Comment:
The code here doesn't work if x, y, and z are not 1D - it would be nice if we allowed x, y and z to be n-dimensional arrays and returned an n-dimensional mark at the end (I guess the easiest might be to unravel)